autogen/oai/github.py

```diff
@@ -0,0 +1,230 @@
+"""Create a Github LLM Client with Azure Fallback.
+
+# Usage example:
+if __name__ == "__main__":
+    config = {
+        "model": "gpt-4o",
+        "system_prompt": "You are a knowledgeable history teacher.",
+        "use_azure_fallback": True
+    }
+
+    wrapper = GithubWrapper(config_list=[config])
+
+    response = wrapper.create(messages=[{"role": "user", "content": "What is the capital of France?"}])
+    print(wrapper.message_retrieval(response)[0])
+
+    conversation = [
+        {"role": "user", "content": "Tell me about the French Revolution."},
+        {"role": "assistant", "content": "The French Revolution was a period of major social and political upheaval in France that began in 1789 with the Storming of the Bastille and ended in the late 1790s with the ascent of Napoleon Bonaparte."},
+        {"role": "user", "content": "What were the main causes?"}
+    ]
+
+    response = wrapper.create(messages=conversation)
+    print(wrapper.message_retrieval(response)[0])
+"""
+
+from __future__ import annotations
+
+import json
+import logging
+import os
+import time
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import requests
+from openai.types.chat import ChatCompletion, ChatCompletionMessage
+from openai.types.chat.chat_completion import Choice
+from openai.types.completion_usage import CompletionUsage
+
+from autogen.cache import Cache
+from autogen.oai.client_utils import should_hide_tools, validate_parameter
+
+logger = logging.getLogger(__name__)
+
+
+class GithubClient:
+    """GitHub LLM Client with Azure Fallback"""
+
+    SUPPORTED_MODELS = [
+        "AI21-Jamba-Instruct",
+        "cohere-command-r",
+        "cohere-command-r-plus",
+        "cohere-embed-v3-english",
+        "cohere-embed-v3-multilingual",
+        "meta-llama-3-70b-instruct",
+        "meta-llama-3-8b-instruct",
+        "meta-llama-3.1-405b-instruct",
+        "meta-llama-3.1-70b-instruct",
+        "meta-llama-3.1-8b-instruct",
+        "mistral-large",
+        "mistral-large-2407",
+        "mistral-nemo",
+        "mistral-small",
+        "gpt-4o",
+        "gpt-4o-mini",
+        "phi-3-medium-instruct-128k",
+        "phi-3-medium-instruct-4k",
+        "phi-3-mini-instruct-128k",
+        "phi-3-mini-instruct-4k",
+        "phi-3-small-instruct-128k",
+        "phi-3-small-instruct-8k",
+    ]
+
+    def __init__(self, **kwargs):
+        self.github_endpoint_url = "https://models.inference.ai.azure.com/chat/completions"
```

Comment:
so here are the supported models and the azure endpoint basically, the difference is you get free (rate limited) inference with a github PAT , and if you want to increase the rate limits you can "fall back" to an azure API key :-)

Reply:
@LittleLittleCloud see here

Reply:
Two questions here
- for falling-back experience, other than providing an azure api key, do you also need to provide another endpoint/resource name/etc. Is the Azure API key the only thing needs to be changed
- I notice that in this PR, you are using openai SDK to make chat completion request, does that work with non-openai models as well
- Continue from Q2, My understand of Github LLM is it's a shim/extra layer on top of azure model-as-service? There is an official SDK (azure.ai.inference) to run inference on azure model-as-service, maybe it's worth adding support for azure.ai.inference in oai client and assistant agent? @sonichi

Reply:
1.  no need to provide any other ressource name or endpoint ,  azure api key does not need to be changed
2. hopefully 
3. works as well as possible with different models, and works fine with openai models
4. it could be smart to rename entirely and pass a flag for optional github key with azure key by default instead ?